core/src/main/java/org/elasticsearch/cluster/metadata/AliasOrIndex.java

```diff
@@ -0,0 +1,133 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import com.google.common.collect.UnmodifiableIterator;
+import org.elasticsearch.common.collect.Tuple;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * If this is an alias it groups all {@link IndexMetaData} instances sharing the same alias name together, otherwise
+ * if this is a concrete index it just encapsulates the {@link IndexMetaData} of that index.
+ */
+public interface AliasOrIndex {
+
+    /**
+     * @return whether this an alias or concrete index
+     */
+    boolean isAlias();
+
+    /**
+     * @return All {@link IndexMetaData} of all concrete indices this alias is referring too or if this is a concrete index its {@link IndexMetaData}
+     */
+    List<IndexMetaData> getIndices();
+
+    class Index implements AliasOrIndex {
+
+        private final IndexMetaData concreteIndex;
+
+        public Index(IndexMetaData indexMetaData) {
+            this.concreteIndex = indexMetaData;
+        }
+
+        @Override
+        public boolean isAlias() {
+            return false;
+        }
+
+        @Override
+        public List<IndexMetaData> getIndices() {
+            return Collections.singletonList(concreteIndex);
+        }
+
+        /**
+         * @return If this is an concrete index, its {@link IndexMetaData}
+         */
+        public IndexMetaData getIndex() {
+            return concreteIndex;
+        }
+
+    }
+
+    class Alias implements AliasOrIndex {
+
+        private final String aliasName;
+        private final List<IndexMetaData> referenceIndexMetaDatas;
+
+        public Alias(AliasMetaData aliasMetaData, IndexMetaData indexMetaData) {
+            this.aliasName = aliasMetaData.getAlias();
+            this.referenceIndexMetaDatas = new ArrayList<>();
+            this.referenceIndexMetaDatas.add(indexMetaData);
+        }
+
+        @Override
+        public boolean isAlias() {
+            return true;
+        }
+
+        @Override
+        public List<IndexMetaData> getIndices() {
+            return referenceIndexMetaDatas;
+        }
+
+        /**
+         * Returns the unique alias metadata per concrete index.
+         *
+         * (note that although alias can point to the same concrete indices, each alias reference may have its own routing
+         * and filters)
+         */
+        public Iterable<Tuple<String, AliasMetaData>> getConcreteIndexAndAliasMetaDatas() {
+            return new Iterable<Tuple<String, AliasMetaData>>() {
```

Comment:
To avoid creating an unnecessary map. It is only iterated once and otherwise a new map needs to be created, which just gets iterated once. Maybe there is a better helper method for this?